compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategiesGPU.cpp

```diff
@@ -266,13 +296,132 @@ void ReductionStrategyStagedThreadDistribution::compute(
+  int64_t getNumThreadsZInBlock() const { return getNumThreadsInBlock()[2]; }
+  std::array<int64_t, 3> getNumThreadsInBlock() const override {
+    std::array<int64_t, 3> res{1, 1, 1};
+    for (int64_t i = 0, e = workgroupTileSizes.size(); i < e; ++i)
+      res[i] = workgroupTileSizes[i];
+    return res;
+  }
+
+  bool isValid() override { return valid; }
+
+ private:
+  bool valid = false;
+
+  /// Compute the small strategy based on the problem size and the
+  /// `maxNumThreadsToUse`.
+  /// `hasTrailingElementwise` is currently used to guard against pathological
+  /// cases where IREE can't bound a buffer and crashes.
+  // TODO: Fix IREE's odegen/Common/PadDynamicAlloc.cpp.
+  void compute(int64_t maxNumThreadsToUse, bool hasTrailingElementwise);
+};
+
+void SmallReductionStrategy::compute(int64_t maxNumThreadsToUse,
+                                     bool hasTrailingElementwise) {
+  assert(maxNumThreadsToUse > 0 && "maxNumThreadsToUse must be > 0");
+  assert(maxNumThreadsToUse >= iree_compiler::kCudaWarpSize &&
+         "not even a warp?");
+
+  // Block-level
+  // ===========
+  FailureOr<int64_t> maybeDivisor =
+      mlir::iree_compiler::maxDivisorOfValueBelowLimit(
+          captures.mostMinorParallelDimensionSize, maxNumThreadsToUse);
+
+  // Trailing elementwise unaligned tiling created bounded local buffers that
+  // IREE is currently incapable of bounding. This results in a crash in
+  // Codegen/Common/PadDynamicAlloc.cpp.
+  // TODO: Capture static parallel dimensions and allow if workgroupTileSizes
+  // divides the parallel dimension evenly.
+  // TODO: More generally fix IREE's PadDynamicAlloc.
+  if (failed(maybeDivisor) && hasTrailingElementwise) return;
+
+  // Dynamic reductions are never supported by default because we can never
+  // know offhand whether we are in a small-reduction regime mode. Since this
+  // mode does not coalesce reads, perf will suffer catastrophically on larger
+  // runtime reduction.
+  // TODO: explicit hint from above that we really want to do that.
+  if (captures.reductionDimensionSize <= 0) return;
+
+  // Otherwise, still only support the small cases for now and fall back to
+  // other strategies otherwise.
+  if (captures.reductionDimensionSize > 2 * iree_compiler::kCudaWarpSize)
+    return;
+
+  // If the captured dimension has no satisfactory divisor, just tile the last
+  // parallel dimension by 2 * iree_compiler::kCudaWarpSize.
+  int64_t numParallelLoops = captures.reductionRank - 1;
+  workgroupTileSizes.append(numParallelLoops, 1);
+  workgroupTileSizes.back() =
+      hasTrailingElementwise
+          ? *maybeDivisor
+          : std::min((int64_t)maxNumThreadsToUse,
+                     (int64_t)(2 * iree_compiler::kCudaWarpSize));
+
+  // Thread-level
+  // ============
+  // Just running sequentially on each thread and relying on cache for
+  // locality.
+  //
+  // Hooray we're good to go.
+  valid = true;
```

Comment:
It feels that these checks should become part of the matcher. A dimension being static and being less than X were the primary use cases for them.

Reply:
Happy to evolve towards that but there are a few caveats right now:
  1. the first check is due to a compiler bug, that should be fixed and should not pollute the matchers.
  2. the other 2 are performance-related constraints for which we lack infra.

I'd prefer to avoid mixing constraints related to structural checks and capture of the IR (what we have today)
with additional constraints related to selecting the strategy and its parameters, which are also dependent on characteristics of the HW.

It will be great to use a similar mechanism for everything but I don't think we have that spelled out properly yet.

My intuition is your comment is more about setting up the stage towards that evolution rather than requesting a change right now?

Reply:
will add TODOs

Reply:
Yes, I saw the compiler bug, that is not a problem. I find it really, really unfortunate that this adds a second mechanism for essentially the same thing as matching. It is surprising when something fails to even start applying even though the match was successful, and works against our argumentation for matchers. What's worse, it's connected to the logic for declaring the strategy _invalid_ instead of _not profitable_.

Now, I understand that we may want to separate the matching for validity from profitability considerations. We should do this cleanly and in a way that can be morphed into dialect constructs later on.

Reply:
> I find it really, really unfortunate that this adds a second mechanism for essentially the same thing as matching.

Do you prefer to wait for a second track of profitability matchers to be spawned off before landing thing?

> What's worse, it's connected to the logic for declaring the strategy invalid instead of not profitable.

Here I fixed it for you: `s/valid/profitable`

Reply:
> I find it really, really unfortunate that this adds a second mechanism for essentially the same thing as matching. 

It is not matching, it is filtering, we do not have a good mechanism for that atm.

> Now, I understand that we may want to separate the matching for validity from profitability considerations. We should do this cleanly and in a way that can be morphed into dialect constructs later on.

Yes, as everything in this line of work, a need is identified and refactored until it becomes cleanly separated and pleasant to use. Everything has followed this M.O. because we decided to bias to action and use C++ rather than wait x years until we have a variadic dialect with constraints, filters, macros and all future bells and whistles that we will discover on the way along.

I am really unclear from your review if you're suggesting to shelve this until some undetermined future time?

Reply:
I think it is crucial to strike the right balance between quick action with lots of shortcuts and forward-looking design work. The minimum bar to pass for me is to avoid misleading others and our future selves about the intended approach. Proper naming, however annoying it may feel, is part of that. So is not engaging too much into a direction that is contrary to the current design decisions (e.g., expressing this in a dialect) without rethinking those decisions. Two weeks from now, we would be reading this code and thinking we wanted an extra "validation" mechanism that is purely C++ and has nothing to do with matching. The more shortcuts take and design contradictions we add, the longer we will have to wait for the variadic dialect with constraints, filters and macros to actually materialize and replace all uses. Conversely, the more we think about that dialect now, the easier it will be to create.

That being said, I did not block this PR, just commented.